packages/server-side-render/src/server-side-render.js

```diff
@@ -83,17 +83,17 @@ export class ServerSideRender extends Component {
 
 	render() {
 		const response = this.state.response;
-		const { className } = this.props;
+		const { className, onEmptyResponse, onErrorResponse, onNullResponse } = this.props;
 		if ( response === '' ) {
-			return (
+			return !! onEmptyResponse ? onEmptyResponse( this ) : (
```

Comment:
Can onEmptyResponse be just a react element, instead of a function ? e.g: !! `onEmptyResponse ? onEmptyResponse : (`

Reply:
In the RSS Block we want to disable the block settings, if we get an error from the SSR. So I think a function is more flexible.

Reply:
I also thought function would be more flexible but open to change if it's required to get this through. Worth mentioning, the code using this component has no access to this components state without a function (passing `this`) which may be important, or not :)

Reply:
The case @Soean provided is interesting:
> In the RSS Block we want to disable the block settings.

We call the function onEmptyResponse on the render phase, render can not have side effects. So, for example, if there is a need to change the local state when the response is empty, that's not possible.
I guess to disable block settings on empty response, we would need to change a local state flag, so this change would not fit the RSS block use case.
I guess needing side effects when the response is empty will be common.
Maybe a possibility that would fit the use case would be an onEmptyResponse event. This function would be called, after the fetch is complete inside the fetch function, with an event object we define as a parameter. As a response to this event, users can apply any side effect desired e.g.: change state. This function would not render anything.
For the render, we have some possibilities:
	- have a prop emptyResponse that is rendered when the response is empty.
	- render nothing if onEmptyResponse is passed, the parent as a response to this event would be responsible for rendering the UI needed.

Any thoughts on this @mikejolley, @Soean?

Reply:
I've no thoughts on this one as my use case didn't require changing state, just acting on the state the block already had.

Reply:
In looking at the comments here I like some of what @jorgefilipecosta proposes.  Here's what I'd like to see that I think will satisfy potential requirements for handling side-effects and a custom empty response render

-  Instead of an empty response event callback fired on completed fetch (as suggested by Jorge), I think it may be more extensible to simply fire any `fetchComplete` callback that is provided in props.  This opens up the possibility for components to trigger side-effects on various fetch states (empty response, successful response, caught fetch error).  This could be available for more advanced use-cases.
-  Just accept render props for the different response state props (and since these are placeholders, let's explicitly call them that) `EmptyResponsePlaceholder`, `NullResponsePlaceholder` and `ErrorResponsePlaceholder` options here and feed any state to them as they are rendererd.  You can then have default values for the render props (which could simplify some of the conditionals here).